File: py/98.验证二叉搜索树.py
#
# @lc app=leetcode.cn id=98 lang=python3
#
# [98] 验证二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # 迭代算法
    # * 使用数组保存中序遍历的结果，在遍历数组判断数组是不是升序排列
    # 52ms   86%   7%   16.5MB
    def isValidBST1(self, root: TreeNode) -> bool:
        travel_res = []
        def dfs(node):
            if node is None:
                return
            dfs(node.left)
            travel_res.append(node.val)
            dfs(node.right)
        dfs(root)
        if len(travel_res) <= 1:
            return True
        for i in range(1,len(travel_res)):
            if travel_res[i]<= travel_res[i-1]:
                return False
        return True

    #  迭代算法
    # 中途就可以判断是不是二叉排序树，不用遍历整个书，跟方法一对比也不用保存全部节点
# 实际效果看起来，Python递归比迭代效率高
# 196ms    6%    33%   16MB
    def isValidBST2(self, root: TreeNode) -> bool:
        travel_res = []
        inorder = float('-inf')
        while root is not None or len(travel_res) != 0:
            while root is not None:
                travel_res.append(root)
                root = root.left
            root = travel_res.pop()
            if root.val <= inorder:
                return False
            inorder = root.val
            root = root.right

        return True

# 递归算法比递归算法1更好 但是从效率上没看出来
# 60ms   49%   42%    15.8MB
    def isValidBST(self, root: TreeNode) -> bool:
        travel_res = []
        def dfs(node,lower,upper):
            if node is None:
                return True
            val = node.val
            if(lower is not None and val >= lower):
                return False
            if(upper is not None and val <= upper):
                return False
            if not dfs(node.left,val,upper):
                return False
            if not dfs(node.right,lower,val):
                return False
            return True
        return dfs(root,None,None)

# @lc code=end

File: java/98.验证二叉搜索树.java
/*
 * @lc app=leetcode.cn id=98 lang=java
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private List<Integer> travelResult = new LinkedList<>();

    /**
     * 使用数组保存中序遍历的结果，在遍历数组判断数组是不是升序排列
     * 109ms   5%    38.5MB
     */
    public boolean isValidBST1(TreeNode root) {
        // if(root == null) return false;
        inOrderDfs(root);
        // if(travelResult.size() == 0 ) return false;
        if(travelResult.size() <= 1 ) return true;
        for(int i=1;i<travelResult.size();i++){
            if(travelResult.get(i-1)>= travelResult.get(i)) return false;
        }
        return true;
    }

    private void inOrderDfs(TreeNode node){
        if(node == null)return;
        inOrderDfs(node.left);
        travelResult.add(node.val);
        inOrderDfs(node.right);
    }


    /**
     * 迭代判断
     * 中途就可以判断是不是二叉排序树，不用遍历整个书，跟方法一对比也不用保存全部节点
     * 2ms    31%     81%   38.3MB
     */
    public boolean isValidBST2(TreeNode root) {
        // if(root == null) return false;
        double inOrder = -Double.MAX_VALUE;
        Deque<TreeNode> stack = new LinkedList<>();
        while(root != null || !stack.isEmpty()){
            while(root != null){
                stack.push(root);
                root = root.left;
            }
            root = stack.pop();
            if(root.val <= inOrder ){
                return false;
            }
            inOrder = root.val;
            root = root.right;
        }
        return true;
    }


    /**
     * 递归算法
     * 减枝算法
     * 1ms    35%     85%    38.2MB
     */
    public boolean isValidBST(TreeNode root) {
        return helper(root, null,null);
    }
    private boolean helper(TreeNode node, Integer lower,Integer upper){
        if(node ==null) return true;

        int val = node.val;
        if( lower!=null && val >= lower){
            return false;
        }

        if(upper!=null && val <= upper){
            return false;
        }

        if(!helper(node.left,val,upper)){
            return false;
        }

        if(!helper(node.right,lower,val)){
            return false;
        }
        return true;
    }
}
// @lc code=end

